dynawo/sources/Models/CPP/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp

```diff
@@ -0,0 +1,579 @@
+    }
+    if (connectionState1 != connectionState1_prev) {
+      setConnected(connectionState1);
+      connectionState1_prev = connectionState1;
+    }
+    if (connectionState2 != connectionState2_prev) {
+      setConnected(connectionState2);
+      connectionState2_prev = connectionState2;
+    }
+    if (gLocal_[0] == ROOT_UP) {
+      UMaxPuReached_ = true;
+      UMinPuReached_ = false;
+      return;
+    }
+    if (gLocal_[1] == ROOT_UP) {
+      UMaxPuReached_ = false;
+      UMinPuReached_ = true;
+      return;
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalJt(const double& /*t*/, const double& cj, SparseMatrix& jt, const int& rowOffset) {
+    if (getIsInitProcess())
+      return;
+    if (!isConnected()) {
+      jt.changeCol();
+      jt.addTerm(UfRawYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(UfYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(IrYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(IiYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(PYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(QYNum_ + rowOffset, 1);
+    } else {
+      jt.changeCol();
+      jt.addTerm(UfRawYNum_ + rowOffset, 1.0 + cj * Tf_);
+      jt.addTerm(UrYNum_ + rowOffset, -Ur()/U());
+      jt.addTerm(UiYNum_ + rowOffset, -Ui()/U());
+      jt.changeCol();
+      jt.addTerm(UfYNum_ + rowOffset, 1);
+      if (UMaxPuReached_ == false && UMinPuReached_ == false) {
+        jt.addTerm(UfRawYNum_ + rowOffset, -1);
+      }
+      jt.changeCol();
+      jt.addTerm(IrYNum_ + rowOffset, 1);
+      jt.addTerm(PYNum_ + rowOffset, -Ur() / U2());
+      jt.addTerm(QYNum_ + rowOffset, -Ui() / U2());
+      jt.addTerm(UrYNum_ + rowOffset, - (P() * U2() - 2 * Ur() * (P() * Ur() + Q() * Ui())) / (U2() * U2()));
+      jt.addTerm(UiYNum_ + rowOffset, - (Q() * U2() - 2 * Ui() * (P() * Ur() + Q() * Ui())) / (U2() * U2()));
+      jt.changeCol();
+      jt.addTerm(IiYNum_ + rowOffset, 1);
+      jt.addTerm(PYNum_ + rowOffset, -Ui() / U2());
+      jt.addTerm(QYNum_ + rowOffset, Ur() / U2());
+      jt.addTerm(UiYNum_ + rowOffset, - (P() * U2() - 2 * Ui() * (P() * Ui() - Q() * Ur())) / (U2() * U2()));
+      jt.addTerm(UrYNum_ + rowOffset, (Q() *  U2() + 2 * Ur() * (P() * Ui() - Q() * Ur())) / (U2() * U2()));
+      jt.changeCol();
+      jt.addTerm(PYNum_ + rowOffset, 1);
+      jt.addTerm(UrYNum_ + rowOffset, - P_dUr(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UiYNum_ + rowOffset, - P_dUi(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UfYNum_ + rowOffset, - P_dUf(U()));
+      jt.changeCol();
+      jt.addTerm(QYNum_ + rowOffset, 1);
+      jt.addTerm(UrYNum_ + rowOffset, - Q_dUr(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UiYNum_ + rowOffset, - Q_dUi(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UfYNum_ + rowOffset, - Q_dUf(U()));
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalG(const double& /*t*/) {
+    gLocal_[0] = (UfRaw() - UMaxPu_ > 0) ? ROOT_UP : ROOT_DOWN;
```

Comment:
to increase stability:
doubleNotEquals(UfRaw(), UMaxPu_) && UfRaw() > UMaxPu_

Reply:
done